File: cli/terrier.go
package main

// Terrier integration (github.com/sylophi/terrier): an external
// registry of repo paths, merged into the project list when the global
// `terrier` toggle is on. Terrier's stable surface is `terrier ls
// --json` plus the rule that a minor version bump is the compatibility
// signal, so that is all this file consumes. Ported alongside
// main/lib/terrier.ts -- the two engines must produce the same merged
// list or the app and the CLI would disagree about which projects
// exist.
//
// Merge semantics: registry.json wins by path. A repo registered in
// both is an ordinary project (removable); one only terrier knows
// becomes a read-only entry with Source "terrier" and a deterministic
// id, so nothing has to be persisted for the two engines to agree.

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"
)

const terrierBinary = "terrier"

// A wedged terrier must not hang every CLI invocation — the merge runs
// pre-dispatch. Mirror of TERRIER_SPAWN_TIMEOUT_MS in
// main/lib/terrier.ts.
const terrierSpawnTimeout = 10 * time.Second

func terrierOutput(args ...string) ([]byte, error) {
	ctx, cancel := context.WithTimeout(context.Background(), terrierSpawnTimeout)
	defer cancel()
	return exec.CommandContext(ctx, terrierBinary, args...).Output()
}

// The registry-read contract this build understands. Terrier's README:
// "a tool checks the minor version and nothing else" -- a minor bump
// means something a tool could be relying on has changed, so an
// unknown minor deactivates the merge rather than guessing. Mirror of
// TERRIER_SUPPORTED_* in main/lib/terrier.ts.
const (
	terrierSupportedMajor = 0
	terrierSupportedMinor = 1
)

// One row of `terrier ls --json`.
type terrierListing struct {
	Path string `json:"path"`
}

var terrierInstalled = sync.OnceValue(func() bool {
	_, err := exec.LookPath(terrierBinary)
	return err == nil
})

func terrierEnabled(global globalConfig) bool {
	return global.Terrier != nil && *global.Terrier
}

// "" when the binary is missing or the spawn fails.
var terrierVersion = sync.OnceValue(func() string {
	stdout, err := terrierOutput("version")
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(stdout))
})

func terrierCompatible() (ok bool, version string) {
	version = terrierVersion()
	var major, minor int
	if n, _ := fmt.Sscanf(version, "v%d.%d", &major, &minor); n < 2 {
		return false, version
	}
	return major == terrierSupportedMajor && minor == terrierSupportedMinor, version
}

var terrierListings = sync.OnceValues(func() ([]terrierListing, error) {
	stdout, err := terrierOutput("ls", "--json")
	if err != nil {
		return nil, err
	}
	var doc struct {
		Projects []terrierListing `json:"projects"`
	}
	if err := json.Unmarshal(stdout, &doc); err != nil {
		return nil, err
	}
	// Home-expanded and required to be absolute — never resolved
	// against cwd, which differs between the app and a shell, so the
	// two engines could mint different ids for the same relative row.
	// Mirrors the filter in main/lib/terrier.ts's listing read.
	var listings []terrierListing
	for _, t := range doc.Projects {
		path := expandHome(t.Path)
		if path == "" || !filepath.IsAbs(path) {
			continue
		}
		listings = append(listings, terrierListing{Path: path})
	}
	return listings, nil
})

// Deterministic id for a terrier-sourced project: UUID-shaped from
// sha256(path) so the TS engine and this one mint the same id for the
// same path without ever writing it down. Uppercased like every
// CLI-minted id. Mirror of terrierProjectId in main/lib/terrier.ts --
// keep the two byte-for-byte in sync (terrier_test.go pins a vector).
func terrierProjectID(path string) string {
	sum := sha256.Sum256([]byte(path))
	h := strings.ToUpper(hex.EncodeToString(sum[:16]))
	return h[0:8] + "-" + h[8:12] + "-" + h[12:16] + "-" + h[16:20] + "-" + h[20:32]
}

// Why the registry can't be read right now, or nil when it can. The
// one walk of the installed -> compatible -> readable ladder, feeding
// the merge's stderr warning and doctor's finding with the same words
// so the two can never explain the same "off" state differently.
// Assumes the caller already checked terrierEnabled: an off toggle is
// the normal quiet state, not trouble. Memoized -- every input is a
// process-lifetime constant.
type terrierTrouble struct {
	summary, advice string
}

var terrierTroubleFor = sync.OnceValue(func() *terrierTrouble {
	if !terrierInstalled() {
		return &terrierTrouble{
			summary: "enabled in config.json but `terrier` isn't on PATH, so no terrier projects are listed",
			advice:  "Install terrier, or turn the toggle off in the app's Settings.",
		}
	}
	if ok, version := terrierCompatible(); !ok {
		return &terrierTrouble{
			summary: fmt.Sprintf("%s isn't a version this build understands (wants v%d.%d), so no terrier projects are listed",
				describeTerrierVersion(version), terrierSupportedMajor, terrierSupportedMinor),
			advice: "Update " + binaryName + " and terrier to versions that agree.",
		}
	}
	if _, err := terrierListings(); err != nil {
		return &terrierTrouble{
			summary: "`terrier ls --json` failed: " + err.Error(),
			advice:  "Run `terrier ls` by hand to see what it says.",
		}
	}
	return nil
})

// The whole gate in one call: nil listings whenever the toggle is off
// or the registry is unreadable (trouble says which; nil trouble with
// nil listings means simply disabled). Every consumer of the terrier
// registry goes through here so no call site can walk the ladder
// differently.
func activeTerrierListings() ([]terrierListing, *terrierTrouble) {
	if !terrierEnabled(readGlobalConfigHints()) {
		return nil, nil
	}
	if trouble := terrierTroubleFor(); trouble != nil {
		return nil, trouble
	}
	listings, _ := terrierListings()
	return listings, nil
}

// Whether the active terrier registry lists path. False whenever the
// integration is off or unreadable -- callers use this to decide id
// continuity, and "don't know" must act like "no". Mirror of
// terrierHasPath in main/lib/terrier.ts.
func terrierHasPath(path string) bool {
	listings, _ := activeTerrierListings()
	for _, t := range listings {
		if t.Path == path {
			return true
		}
	}
	return false
}

// The pre-dispatch merge (main.go): registry entries as-is, then a
// read-only project per terrier repo the registry doesn't already
// hold. Failures degrade to the registry alone with one stderr note --
// a broken terrier must not take every command down with it.
func mergeTerrierProjects(projects []project) []project {
	listings, trouble := activeTerrierListings()
	if trouble != nil {
		note(yellowErr("warning:") + " " + trouble.summary + ".")
	}
	return appendTerrierProjects(projects, listings)
}

// The pure half of the merge, split out for tests. Ordering must match
// mergeTerrierProjects in main/lib/projects/index.ts: registry order
// first, terrier extras after, sorted by name then path (plain byte
// compare in both engines).
func appendTerrierProjects(projects []project, listings []terrierListing) []project {
	known := make(map[string]bool, len(projects))
	for _, p := range projects {
		known[p.Path] = true
	}
	var extras []project
	for _, t := range listings {
		if t.Path == "" || known[t.Path] {
			continue
		}
		known[t.Path] = true
		extras = append(extras, project{
			ID:     terrierProjectID(t.Path),
			Name:   filepath.Base(t.Path),
			Path:   t.Path,
			Source: "terrier",
		})
	}
	sort.Slice(extras, func(i, j int) bool {
		if extras[i].Name != extras[j].Name {
			return extras[i].Name < extras[j].Name
		}
		return extras[i].Path < extras[j].Path
	})
	return append(projects, extras...)
}

func describeTerrierVersion(version string) string {
	if version == "" {
		return "(version unreadable)"
	}
	return version
}

File: host/lib/terrier.ts
// Terrier integration (github.com/sylophi/terrier): an external
// registry of repo paths, merged into the project list when the global
// `terrier` toggle is on. Terrier's stable surface is `terrier ls
// --json` plus the rule that a minor version bump is the compatibility
// signal, so that is all this file consumes. Ported alongside
// cli/terrier.go: the two engines must produce the same merged list
// or the app and the CLI would disagree about which projects exist.
//
// The registry read is a subprocess, but project resolution
// (findProjectOrThrow) is sync and hot, so sync callers read the
// cache's last-known value via peek() while async ones await a
// refresh. The projects:list handler always refreshes first, and the
// renderer only ever learns a terrier project's id from that list, so
// by the time an id reaches a sync resolver the cache holds it.
import { execFile } from "node:child_process";
import { createHash } from "node:crypto";
import { isAbsolute } from "node:path";
import { promisify } from "node:util";
import { z } from "zod";
import type { TerrierReadiness } from "@shared/schemas";
import { readGlobalConfig } from "./config/global";
import { expandHome } from "./util/paths";
import { ttlValueCache } from "./util/ttlCache";

const execFileP = promisify(execFile);

const TERRIER_BINARY = "terrier";
// A wedged terrier must not hang the callers gated on these spawns:
// projects:list awaits the listing refresh, so an unbounded wait here
// is a sidebar that never loads. Mirror of terrierSpawnTimeout in
// cli/terrier.go.
const TERRIER_SPAWN_TIMEOUT_MS = 10_000;

function execTerrier(args: string[]): Promise<{ stdout: string }> {
  return execFileP(TERRIER_BINARY, args, {
    timeout: TERRIER_SPAWN_TIMEOUT_MS,
  });
}

// The registry-read contract this build understands. Terrier's README
// says a tool checks the minor version and nothing else: a minor bump
// means something a tool could be relying on has changed, so an
// unknown minor deactivates the merge rather than guessing. Mirror of
// terrierSupported* in cli/terrier.go.
const TERRIER_SUPPORTED_MAJOR = 0;
const TERRIER_SUPPORTED_MINOR = 1;

// `terrier ls --json` rows. Local to this module on purpose: it
// describes another tool's output, not this app's IPC contract, so it
// doesn't belong in the shared barrel (same as GhPrListItemSchema in
// githubCli/pullRequests.ts). The path is allowed to be empty here and
// filtered below, so one blank row costs that row rather than the
// whole document, matching how the Go engine reads the same output.
const TerrierListingSchema = z.object({ path: z.string() });
const TerrierLsSchema = z.object({
  projects: z.array(TerrierListingSchema),
});
export type TerrierListing = z.infer<typeof TerrierListingSchema>;

const READINESS_CACHE_TTL_MS = 30_000;
const LIST_CACHE_TTL_MS = 15_000;

// One spawn answers both questions: ENOENT is "not installed", any
// output is the version to run the minor handshake against.
const readinessCache = ttlValueCache<TerrierReadiness>(
  READINESS_CACHE_TTL_MS,
  async () => {
    let version: string;
    try {
      ({ stdout: version } = await execTerrier(["version"]));
    } catch (error) {
      const installed = (error as NodeJS.ErrnoException).code !== "ENOENT";
      return { installed, compatible: false };
    }
    version = version.trim();
    return {
      installed: true,
      compatible: versionCompatible(version),
      version: version || undefined,
    };
  },
);

function versionCompatible(version: string): boolean {
  const match = /^v(\d+)\.(\d+)/.exec(version);
  if (!match) return false;
  return (
    Number(match[1]) === TERRIER_SUPPORTED_MAJOR &&
    Number(match[2]) === TERRIER_SUPPORTED_MINOR
  );
}

export function terrierReadiness(): Promise<TerrierReadiness> {
  return readinessCache.get();
}

// Empty whenever any gate is closed (toggle off, binary missing,
// version handshake failed) or the read fails: a broken terrier must
// degrade to "no terrier projects", never break project loading.
async function fetchListings(): Promise<TerrierListing[]> {
  const global = await readGlobalConfig();
  if (!global.terrier) return [];
  const readiness = await terrierReadiness();
  if (!readiness.installed || !readiness.compatible) return [];
  try {
    const { stdout } = await execTerrier(["ls", "--json"]);
    const parsed = TerrierLsSchema.safeParse(JSON.parse(stdout));
    if (!parsed.success) {
      console.warn("[terrier] unrecognized ls --json shape:", parsed.error);
      return [];
    }
    // Home-expanded and required to be absolute, never resolved
    // against cwd: that differs between the app and a shell, so the
    // two engines could mint different ids for the same relative row.
    // Mirrors the filter in cli/terrier.go's listing read.
    const listings: TerrierListing[] = [];
    for (const row of parsed.data.projects) {
      const path = expandHome(row.path);
      if (path === "" || !isAbsolute(path)) continue;
      listings.push({ path });
    }
    return listings;
  } catch (error) {
    console.warn("[terrier] ls --json failed:", error);
    return [];
  }
}

const listCache = ttlValueCache(LIST_CACHE_TTL_MS, fetchListings);

// fetchListings never throws, so neither does this.
export async function refreshTerrierListings(): Promise<void> {
  await listCache.get();
}

// For the sync callers (loadProjects and everything on top of it):
// the last-known listings, possibly stale until an async caller
// refreshes. Better than absent, which would make every terrier
// project id transiently unresolvable (invalidateTerrierCaches uses
// expire() for exactly this reason).
export function terrierListingsSnapshot(): TerrierListing[] {
  return listCache.peek() ?? [];
}

// Whether the last-known terrier listings hold path. False whenever
// the integration is off or unrefreshed: callers use this to decide
// id continuity, and "don't know" must act like "no". Mirror of
// terrierHasPath in cli/terrier.go.
export function terrierHasPath(path: string): boolean {
  return terrierListingsSnapshot().some((t) => t.path === path);
}

// Whether the project would live on as a terrier-sourced project with
// the same id after its registry entry is dropped: the remove
// handler's cue that nothing is actually going away, so scripts and
// app-side state must be left alone. Mirrors the id-continuity branch
// in cmdProjectRemove (cli/cmd_project.go), where the CLI keeps or
// relocates the state dir on the same rule.
export function terrierRetainsProject(project: {
  id: string;
  path: string;
}): boolean {
  return (
    terrierHasPath(project.path) &&
    terrierProjectId(project.path) === project.id
  );
}

// For writers that change what the caches answer (the global-config
// write flipping the toggle): the next read re-asks instead of serving
// up to a TTL of the pre-write world. Expire, not invalidate: erasing
// the listings would blank the snapshot until the refetch lands, and
// every sync resolver would throw unknownProjectError for terrier
// projects still on screen.
export function invalidateTerrierCaches(): void {
  readinessCache.expire();
  listCache.expire();
}

// Deterministic id for a terrier-sourced project: UUID-shaped from
// sha256(path) so the Go engine and this one mint the same id for the
// same path without ever writing it down. Uppercase like CLI-minted
// ids. Mirror of terrierProjectID in cli/terrier.go. Keep the two
// byte-for-byte in sync (cli/terrier_test.go pins a vector).
export function terrierProjectId(path: string): string {
  const hex = createHash("sha256")
    .update(path)
    .digest("hex")
    .slice(0, 32)
    .toUpperCase();
  return [
    hex.slice(0, 8),
    hex.slice(8, 12),
    hex.slice(12, 16),
    hex.slice(16, 20),
    hex.slice(20, 32),
  ].join("-");
}
